File: modules/benchmarks/src/circles.rs
//! STDB module used for benchmarks based on "realistic" workloads we are focusing in improving.
use crate::Load;
use spacetimedb::{log, ReducerContext, SpacetimeType, Table, Timestamp};
use std::hint::black_box;

#[derive(SpacetimeType, Debug, Clone, Copy)]
pub struct Vector2 {
    pub x: f32,
    pub y: f32,
}

// ---------- schemas ----------

#[spacetimedb::table(name = entity)]
pub struct Entity {
    #[auto_inc]
    #[primary_key]
    pub id: u32,
    pub position: Vector2,
    pub mass: u32,
}

impl Entity {
    pub fn new(id: u32, x: f32, y: f32, mass: u32) -> Self {
        Self {
            id,
            mass,
            position: Vector2 { x, y },
        }
    }
}

#[spacetimedb::table(name = circle)]
pub struct Circle {
    #[primary_key]
    pub entity_id: u32,
    #[index(btree)]
    pub player_id: u32,
    pub direction: Vector2,
    pub magnitude: f32,
    pub last_split_time: Timestamp,
}

impl Circle {
    pub fn new(entity_id: u32, player_id: u32, x: f32, y: f32, magnitude: f32, last_split_time: Timestamp) -> Self {
        Self {
            entity_id,
            player_id,
            direction: Vector2 { x, y },
            magnitude,
            last_split_time,
        }
    }
}

#[spacetimedb::table(name = food)]
pub struct Food {
    #[primary_key]
    pub entity_id: u32,
}

impl Food {
    pub fn new(entity_id: u32) -> Self {
        Self { entity_id }
    }
}

fn mass_to_radius(mass: u32) -> f32 {
    (mass as f32).sqrt()
}

fn is_overlapping(entity1: &Entity, entity2: &Entity) -> bool {
    let entity1_radius = mass_to_radius(entity1.mass);
    let entity2_radius = mass_to_radius(entity2.mass);
    let distance =
        ((entity1.position.x - entity2.position.x).powi(2) + (entity1.position.y - entity2.position.y).powi(2)).sqrt();
    distance < entity1_radius.max(entity2_radius)
}

// ---------- insert bulk ----------
#[spacetimedb::reducer]
pub fn insert_bulk_entity(ctx: &ReducerContext, count: u32) {
    for id in 0..count {
        ctx.db
            .entity()
            .insert(Entity::new(0, id as f32, (id + 5) as f32, id * 5));
    }
    log::info!("INSERT ENTITY: {count}");
}

#[spacetimedb::reducer]
pub fn insert_bulk_circle(ctx: &ReducerContext, count: u32) {
    for id in 0..count {
        ctx.db.circle().insert(Circle::new(
            id,
            id,
            id as f32,
            (id + 5) as f32,
            (id * 5) as f32,
            ctx.timestamp,
        ));
    }
    log::info!("INSERT CIRCLE: {count}");
}

#[spacetimedb::reducer]
pub fn insert_bulk_food(ctx: &ReducerContext, count: u32) {
    for id in 1..=count {
        ctx.db.food().insert(Food::new(id));
    }
    log::info!("INSERT FOOD: {count}");
}

// Simulate
// ```
// SELECT * FROM Circle, Entity, Food
// ```
#[spacetimedb::reducer]
pub fn cross_join_all(ctx: &ReducerContext, expected: u32) {
    let mut count = 0;
    for _circle in ctx.db.circle().iter() {
        for _entity in ctx.db.entity().iter() {
            for _food in ctx.db.food().iter() {
                count += 1;
            }
        }
    }

    log::info!("CROSS JOIN ALL: {expected}, processed: {count}");
}

// Simulate
// ```
// SELECT * FROM Circle JOIN ENTITY USING(entity_id), Food JOIN ENTITY USING(entity_id)
// ```
#[spacetimedb::reducer]
pub fn cross_join_circle_food(ctx: &ReducerContext, expected: u32) {
    let mut count = 0;
    for circle in ctx.db.circle().iter() {
        let Some(circle_entity) = ctx.db.entity().id().find(circle.entity_id) else {
            continue;
        };

        for food in ctx.db.food().iter() {
            count += 1;
            let food_entity = ctx
                .db
                .entity()
                .id()
                .find(food.entity_id)
                .unwrap_or_else(|| panic!("Entity not found: {}", food.entity_id));
            black_box(is_overlapping(&circle_entity, &food_entity));
        }
    }

    log::info!("CROSS JOIN CIRCLE FOOD: {expected}, processed: {count}");
}

#[spacetimedb::reducer]
pub fn init_game_circles(ctx: &ReducerContext, initial_load: u32) {
    let load = Load::new(initial_load);

    insert_bulk_food(ctx, load.initial_load);
    insert_bulk_entity(ctx, load.initial_load);
    insert_bulk_circle(ctx, load.small_table);
}

#[spacetimedb::reducer]
pub fn run_game_circles(ctx: &ReducerContext, initial_load: u32) {
    let load = Load::new(initial_load);

    cross_join_circle_food(ctx, initial_load * load.small_table);
    cross_join_all(ctx, initial_load * initial_load * load.small_table);
}

File: modules/benchmarks-cs/circles.cs
using SpacetimeDB;

namespace Benchmarks;

public static partial class circles
{
    [SpacetimeDB.Type]
    public partial struct Vector2(float x, float y)
    {
        public float x = x;
        public float y = y;
    }

    [SpacetimeDB.Table(Name = "entity")]
    public partial struct Entity(uint id, float x, float y, uint mass)
    {
        [AutoInc]
        [PrimaryKey]
        public uint id = id;
        public Vector2 position = new(x, y);
        public uint mass = mass;
    }

    [SpacetimeDB.Table(Name = "circle")]
    public partial struct Circle(uint entity_id, uint player_id, float x, float y, float magnitude)
    {
        [PrimaryKey]
        public uint entity_id = entity_id;

        [SpacetimeDB.Index.BTree]
        public uint player_id = player_id;
        public Vector2 direction = new(x, y);
        public float magnitude = magnitude;
        public Timestamp last_split_time = (Timestamp)DateTimeOffset.UtcNow;
    }

    [SpacetimeDB.Table(Name = "food")]
    public partial struct Food(uint entity_id)
    {
        [PrimaryKey]
        public uint entity_id = entity_id;
    }

    public static float MassToRadius(uint mass)
    {
        return (float)Math.Sqrt(mass);
    }

    public static bool IsOverlapping(Entity entity1, Entity entity2)
    {
        float entity1_radius = MassToRadius(entity1.mass);
        float entity2_radius = MassToRadius(entity2.mass);
        float distance = (float)
            Math.Sqrt(
                Math.Pow(entity1.position.x - entity2.position.x, 2)
                    + Math.Pow(entity1.position.y - entity2.position.y, 2)
            );
        return distance < Math.Max(entity1_radius, entity2_radius);
    }

    [SpacetimeDB.Reducer]
    public static void insert_bulk_entity(ReducerContext ctx, uint count)
    {
        for (uint id = 0; id < count; id++)
        {
            ctx.Db.entity.Insert(new(0, id, id + 5, id * 5));
        }
        Log.Info($"INSERT ENTITY: {count}");
    }

    [SpacetimeDB.Reducer]
    public static void insert_bulk_circle(ReducerContext ctx, uint count)
    {
        for (uint id = 0; id < count; id++)
        {
            ctx.Db.circle.Insert(new(id, id, id, id + 5, id * 5));
        }
        Log.Info($"INSERT CIRCLE: {count}");
    }

    [SpacetimeDB.Reducer]
    public static void insert_bulk_food(ReducerContext ctx, uint count)
    {
        for (uint id = 1; id <= count; id++)
        {
            ctx.Db.food.Insert(new(id));
        }
        Log.Info($"INSERT FOOD: {count}");
    }

    [SpacetimeDB.Reducer]
    public static void cross_join_all(ReducerContext ctx, uint expected)
    {
        uint count = 0;
        foreach (Circle circle in ctx.Db.circle.Iter())
        {
            foreach (Entity entity in ctx.Db.entity.Iter())
            {
                foreach (Food food in ctx.Db.food.Iter())
                {
                    count++;
                }
            }
        }

        Log.Info($"CROSS JOIN ALL: {expected}, processed: {count}");
    }

    [SpacetimeDB.Reducer]
    public static void cross_join_circle_food(ReducerContext ctx, uint expected)
    {
        uint count = 0;
        foreach (Circle circle in ctx.Db.circle.Iter())
        {
            if (ctx.Db.entity.id.Find(circle.entity_id) is not { } circle_entity)
            {
                continue;
            }

            foreach (Food food in ctx.Db.food.Iter())
            {
                count++;
                Entity food_entity =
                    ctx.Db.entity.id.Find(food.entity_id)
                    ?? throw new Exception($"Entity not found: {food.entity_id}");
                Bench.BlackBox(IsOverlapping(circle_entity, food_entity));
            }
        }

        Log.Info($"CROSS JOIN CIRCLE FOOD: {expected}, processed: {count}");
    }

    [SpacetimeDB.Reducer]
    public static void init_game_circles(ReducerContext ctx, uint initial_load)
    {
        Load load = new(initial_load);

        insert_bulk_food(ctx, load.initial_load);
        insert_bulk_entity(ctx, load.initial_load);
        insert_bulk_circle(ctx, load.small_table);
    }

    [SpacetimeDB.Reducer]
    public static void run_game_circles(ReducerContext ctx, uint initial_load)
    {
        Load load = new(initial_load);

        cross_join_circle_food(ctx, initial_load * load.small_table);
        cross_join_all(ctx, initial_load * initial_load * load.small_table);
    }
}
